docs/tutorials/PyTorch.py

```diff
@@ -0,0 +1,322 @@
+#
+# We can see if PyTorch gets the same answer as us if we do:
+
+y.backward()  # populates gradient (.grad) attributes of y with respect to all of its independent variables
+x.grad  # returns the grad attribute (the gradient) of y with respect to x
+
+# PyTorch uses the concept of automatic differentiation to calculate the derivative. Instead of computing the derivative as we would by hand, the program is using a computational graph.
+
+# ## Optimizing a Function with Gradient Descent
+#
+# If we were on the side of a hill in the dark and we wanted to get down to the bottom of a valley, how would we do it?
+#
+# We wouldn't be able to see all the way to the bottom of the valley, but we could feel which way is down based on where we are standing. We would take steps in the downward direction and we'd know when to stop when the ground felt flat.
+#
+# One other thing we'd have to consider is our step size. If we take very small steps in the direction of the descent, it will take us a longer time than if we take larger steps. However, if we take super long steps, we might completely miss the flat part of the valley, and start ascending the other side of the valley.
+
+# We can look at the gradient descent from a more mathematical lense by looking at the graph z = t ** 2:
+
+t = np.linspace(-5, 5, 100)
+plt.plot(t, t ** 2)
+
+# We start on the left hill at the point (-4, 16):
+
+t = np.linspace(-5, 5, 100)
+plt.plot(t, t ** 2)
+plt.scatter(-4, 16)
+plt.show()
+
+# If we plot a tangent line at this point, we see which directions we can go:
+
+# +
+t = np.linspace(-5, 5, 100)
+plt.plot(t, t ** 2)
+
+
+plt.scatter(-4, 16)  # point
+tp = -4
+yp = 16
+
+plt.plot(t, tp * 2 * (t - tp) + yp)  # tangent line
+
+plt.xlim(xmin=-5, xmax=5)
+plt.ylim(ymin=0, ymax=25)
+plt.show()
+# -
+
+# We see we need to go down to go toward the minimum. We take a very small step in the direction of the steepest downward slope. When we take steps, we find the x coordinate of our new location by this equation:
+#
+# $x_{new} = x_{current} - \nabla y(x_{current}) * (step \: size)$
+#
+# where:
+# - $x_{current}$ is our current x value
+#
+# - $\nabla y(x_{current})$ is the gradient at our current point
+#
+# - $(step \: size)$ is a value we choose that tells us scales our steps
+#
+# This makes sense because we start at some x value and we want to get to a new x value that is closer to the minimum. In the above plot, the gradient is negative. If we were to add the gradient to our current x value that would bring us further away from the minimum to a more negative x value. This is because the gradient points in the direction of the steepest ascent while we are looking to go in the direction of the steepest descent. This is why there is a negative in the equation.
+#
+#
+# We will choose `step size = 0.1`:
+
+# +
+t = np.linspace(-5, 5, 100)
+plt.plot(t, t ** 2)
+
+
+plt.scatter(-4, 16)  # point
+
+plt.scatter(-3.2, 10.24)  # second point
```

Comment:
How are the values computed in these points? Would it be possible to compute them directly, rather than hard-coding them?